python/lsst/daf/butler/_butler.py

```diff
@@ -0,0 +1,1066 @@
+# for details of code ownership.
+#
+# This software is dual licensed under the GNU General Public License and also
+# under a 3-clause BSD license. Recipients may choose which of these licenses
+# to use; please see the files gpl-3.0.txt and/or bsd_license.txt,
+# respectively.  If you choose the GPL option then the following text applies
+# (but note that there is still no warranty even if you opt for BSD instead):
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+
+from __future__ import annotations
+
+from abc import abstractmethod
+from collections.abc import Collection, Iterable, Sequence
+from contextlib import AbstractContextManager
+from typing import Any, TextIO
+
+from lsst.resources import ResourcePath, ResourcePathExpression
+from lsst.utils import doImportType
+from lsst.utils.logging import getLogger
+
+from ._butler_config import ButlerConfig
+from ._butler_repo_index import ButlerRepoIndex
+from ._config import Config, ConfigSubset
+from ._dataset_existence import DatasetExistence
+from ._dataset_ref import DatasetIdGenEnum, DatasetRef
+from ._dataset_type import DatasetType
+from ._deferredDatasetHandle import DeferredDatasetHandle
+from ._file_dataset import FileDataset
+from ._limited_butler import LimitedButler
+from ._storage_class import StorageClass
+from .datastore import DatasetRefURIs, Datastore
+from .dimensions import DataId, DimensionConfig
+from .registry import Registry, RegistryConfig, _RegistryFactory
+from .repo_relocation import BUTLER_ROOT_TAG
+from .transfers import RepoExportContext
+
+log = getLogger(__name__)
+
+
+class Butler(LimitedButler):
+    """Interface for data butler and factory for Butler instances.
+
+    Parameters
+    ----------
+    config : `ButlerConfig`, `Config` or `str`, optional.
+        Configuration. Anything acceptable to the `ButlerConfig` constructor.
+        If a directory path is given the configuration will be read from a
+        ``butler.yaml`` file in that location. If `None` is given default
+        values will be used. If ``config`` contains "cls" key then its value is
+        used as a name of butler class and it must be a sub-class of this
+        class, otherwise `DirectButler` is instantiated.
+    **kwargs : `Any`
+        Optional keyword arguments passed to a constructor of actual butler
+        class.
```

Comment:
OK, I added all those parameters to `Butler(...)`  and `Butler.from_config(...)`. I also remembered that I intended to move examples from `DirectButler` docstring to `Butler.from_config`, so that is also on a separate commit with these new parameters.